src/h2load_http1_session.cc

```diff
@@ -0,0 +1,160 @@
+/*
+ * nghttp2 - HTTP/2 C Library
+ *
+ * Copyright (c) 2015 British Broadcasting Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "h2load_http1_session.h"
+
+#include <cassert>
+#include <cerrno>
+
+#include "h2load.h"
+#include "util.h"
+#include "template.h"
+
+#include <iostream>
+#include <fstream>
+
+#include "http-parser/http_parser.h"
+
+using namespace nghttp2;
+
+namespace h2load {
+
+Http1Session::Http1Session(Client *client)
+    : stream_req_counter_(1), stream_resp_counter_(1), client_(client),
+      htp_(make_unique<http_parser>()), complete_(false) {
+  http_parser_init(htp_.get(), HTTP_RESPONSE);
+  htp_->data = this;
+}
+
+Http1Session::~Http1Session() {}
+
+namespace {
+// HTTP response message begin
+int htp_msg_begincb(http_parser *htp) { return 0; }
+} // namespace
+
+namespace {
+// HTTP response status code
+int htp_statuscb(http_parser *htp, const char *at, size_t length) {
+  auto session = static_cast<Http1Session *>(htp->data);
+  auto client = session->get_client();
+  client->on_status_code(session->stream_resp_counter_, htp->status_code);
+
+  return 0;
+}
+} // namespace
+
+namespace {
+// HTTP response message complete
+int htp_msg_completecb(http_parser *htp) {
+  auto session = static_cast<Http1Session *>(htp->data);
+  auto client = session->get_client();
+
+  client->on_stream_close(session->stream_resp_counter_, true,
+                          session->req_stats_[session->stream_resp_counter_]);
+
+  session->stream_resp_counter_ += 2;
+
+  return 0;
+}
+} // namespace
+
+namespace {
+int htp_body_cb(http_parser *htp, const char *data, size_t len) {
+  auto session = static_cast<Http1Session *>(htp->data);
+  auto client = session->get_client();
+
+  client->record_ttfb();
+
```

Comment:
I think we need to update response body bytes received here, just like we do in h2, spdy:
```
client->worker->stats.bytes_body += len;
```